### ISSUE TITLE
Backprop: error increasing ?

### ISSUE THREAD

**Issue by stephane-lallee** [accord-net/framework#105]
Hi !

I am using a simple MLP with the Parrallel Resilient Backpropagation.

I noticed a strange behavior on some training set using the batch training : the error first goes down rapidly and then it... increases. How is that possible ? I would understand that it stay stable to a high level, but how can it increase?

I would really appreciate any help regarding this...

Cheers,
Stephane

**Comment by cesarsouza** [comment]
Hi Stephane,

Resilient backpropagation doesn't really take the gradient into account, but instead relies in some heuristics to control how much a weight should change between two iterations. If your batches are different, the error rate computed after running a batch may indeed increase when the model is run on another batch.

However, this might as well be a bug I hadn't noticed before. Could you check if the issue also happens with the non-parallel version of the algorithm?

Thanks for reporting this issue!

Best regards,
Cesar

**Comment by stephane-lallee** [comment]
Hi Cesar,

The same behavior happens in both cases (parallel or not).
I'm of course talking about multiple epochs on the same dataset, not different dataset.
Interestingly this problem happens only on some specific datasets, some other are perfectly fine and the error keep decreasing no matter how long I run the algorithm...

Now that I plot the error instead of the printing it out it looks quite "stable" (although it is increasing).
Is this the intended behavior? Also the error is quite high... and when is use the network, even with a high error the prediction are quite accurate. Could it be that something is wrong (or that I did not understand) in the error reporting?

Here is a snippet of the code I use:
```C#
var function = new SigmoidFunction(2.0);
ActivationNetwork goalNetwork = goalNetwork = new ActivationNetwork(function, 2 * maze.StatesCount, 20, maze.StatesCount);
ParallelResilientBackpropagationLearning goalTeacher = new ParallelResilientBackpropagationLearning(goalNetwork);
goalTeacher.Reset(0.0125);

StreamWriter test = new StreamWriter("test.txt");
double error = double.PositiveInfinity;
for (epoch = 0; epoch < 5000 && error > stopError; epoch++)
{
     error = goalTeacher.RunEpoch(input, output);
     test.WriteLine("Epoch " + epoch + " = \t" + error)
}
test.Close();
```C#

Here are plots of the result:
![image](https://cloud.githubusercontent.com/assets/6750632/7974136/a49daf46-0a90-11e5-8de1-6ad82e2aa5ac.png)

And if we omit the first steps:
![image](https://cloud.githubusercontent.com/assets/6750632/7974153/e3a065da-0a90-11e5-8bbb-b1e1caf3fe3b.png)

Thanks again for your reply !

**Comment by cesarsouza** [comment]
Hi Stephane,

Your code seems perfect. The point is that the Resilient Backpropagation algorithm doesn't use the gradient itself to make weight updates, and as such, perhaps it might not even be guaranteed to converge to a minimum. In fact, it is an heuristic algorithm in nature. I will surely double-check the code, but I would expect this behavior more or less expected. 

By the way, the error is quite high because by default the error is reported as the total sum of squared errors, and not as the average error as it would be expected. If you divide your error by the number of training samples, then you would get the mean squared error instead.

**Comment by stephane-lallee** [comment]
I see. I did not read in detail the algo of the resilient backprop. It appeared so much faster that the original one that I just did not even check. I'll do some more comparison and choose the most adapted one :-)
Thanks for the tip ! I close the issue ! (and by the way, congratulation for this framework, it is really really nice...)

**Comment by cesarsouza** [comment]
Thanks! Please let me know of your findings  :smiley:

I will keep the issue open for some time in case someone else comes across this thread and can help explain why this behavior would occur. I will also investigate it in the future, once I have some free time. Thanks again for reporting the issue!

Regards,
Cesar
